### ISSUE TITLE
Deadlock in latest 3.4.0

### ISSUE THREAD

**Issue by angelix** [dizitart/nitrite-database#212]
In the recent release (3.4.0), my app is hanging in an upsert update (ObjectRepository)
Maybe is related to the changes made in #193.

**Comment by anidotnet** [comment]
If you are using any locking mechanism to synchronize nitrite write, remove it. If not, please share a reproducible code.

**Comment by angelix** [comment]
It seems removing a lock from an outer method, fixed the issue, the locks where there from an implementation before using nitrite.
But why is this happening? How an app lock can mess the sync in nitrite?

**Comment by anidotnet** [comment]
In version 3.4.0 locking mechanism of nitrite has changed from 3.3.0. In 3.3.0 app developer needs to synchronize read-write and indexing separately, now from 3.4.0 it is managed by nitrite only. For upsert, update/insert and index write all are managed using a single `ReentrantReadWriteLock` in nitrite.

**Comment by angelix** [comment]
I was able to reproduce it. Try using 2 keys in an upsert update.

eg.
```
collection.update(Filters.and(eq("key", "key"), eq("second_key", "second_key")), doc, UpdateOptions.updateOptions(true));
```

If you need i can share the code, but its very easy reproducible.

I couldn't pass the invokeAll method in debug. (AndFilters.java)
```
List<Callable<Set<NitriteId>>> tasks = createTasks(filters, documentMap);

            boolean initialCount = true;
            List<Future<Set<NitriteId>>> futures = executorService.invokeAll(tasks);
```

**Comment by anidotnet** [comment]
Did you run the above code with 3.4.0 or 3.3.0? I ran below test code on 3.4.0 and it went fine without any issues

```java
@Test
public void testIssue212() {
    NitriteCollection collection = db.getCollection("test");
    Document doc1 = createDocument("key", "key").put("second_key", "second_key").put("third_key", "third_key");
    Document doc2 = createDocument("key", "key").put("second_key", "second_key").put("fourth_key", "fourth_key");
    Document doc = createDocument("fifth_key", "fifth_key");

    collection.insert(doc1, doc2);
    collection.update(Filters.and(Filters.eq("key", "key"),
        Filters.eq("second_key", "second_key")), doc, UpdateOptions.updateOptions(true));

    for (Document document : collection.find()) {
        System.out.println(document);
    }
}
```

**Comment by angelix** [comment]
Try adding indexes for both fields. Sorry, i didn't mention it.

**Comment by anidotnet** [comment]
Nice find.. I am going to give an emergency fixed version in 3.4.1
